Fix relaunching scancode on Apple silicon using Rosetta 2 emulation #2835

MarcelBochtler opened the pull request:
<!-- Delete Template sections if unneccesary -->
<!-- Add issue number here (We encourage you to create the Issue First) -->
<!-- You can also link the issue in Commit Messages -->

Related to #2835.

### Tasks

* [x] Reviewed [contribution guidelines](https://github.com/nexB/scancode-toolkit/blob/develop/CONTRIBUTING.rst)
* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [x] Tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR
  Run [tests](https://scancode-toolkit.readthedocs.io/en/latest/contribute/contrib_dev.html#running-tests) locally to check for errors. 
* [x] Commits are in uniquely-named feature branch and has no merge conflicts 📁

<!--
We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged**
if your tests fail at first!
If tests do fail, click on the red `X` to learn why by reading the logs.
Thanks!
-->

<!-- Don't forget to Signoff -->

MarcelBochtler committed:
Only relaunch scancode when not in emulation mode #2835

Running `uname -m` on Apple silicon machines always results in `arm64`,
regardless if Bash is running in Rosetta 2 emulation or natively.
Check if scancode and configure is running in emulation mode (see [1]),
before relaunching it, to avoid an infinite loop.

[1]: https://developer.apple.com/documentation/apple-silicon/about-the-rosetta-translation-environment

Signed-off-by: Marcel Bochtler <git@bochtler.io>

MarcelBochtler committed:
Pass all parameters to the relaunched shell

Calling `bash -c` with `$@` results in only the first parameter being
passed to the subshell. Use `$*` to ensure all parameters are correctly
passed to the Rosetta 2 emulated subshell.

Signed-off-by: Marcel Bochtler <git@bochtler.io>

pombredanne reviewed (approved):
@MarcelBochtler Thank you ++ for testing and fixing this.... you are awesome!

pombredanne committed:
Support extractcode on Apple ARM chip nexB#2835

This commit carries over a patch applied to the scancode and configure
scripts to ensure that extractcode can work correctly in X86 emulated
mode on Apple ARM silicon,

* Only relaunch extractcode when not in emulation mode
Running `uname -m` on Apple silicon machines always results in `arm64`,
regardless if Bash is running in Rosetta 2 emulation or natively. Check
if extractcode is running in emulation mode (see [1]), before
relaunching it, to avoid an infinite loop.

*Pass all parameters to the relaunched shell
Calling `bash -c` with `$@` results in only the first parameter being
passed to the subshell. Use `$*` to ensure all parameters are correctly
passed to the Rosetta 2 emulated subshell.

[1]: https://developer.apple.com/documentation/apple-silicon/about-the-rosetta-translation-environment

Reported-by: Marcel Bochtler <git@bochtler.io>
Authored-by: Marcel Bochtler <git@bochtler.io>
Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne commented:
@MarcelBochtler I carried over your patch to the `extractcode` script too... And I will merge shortly.
Thank you ++ again :bow:

pombredanne commented:
Thanks!
Merging
